services/comments.js

```diff
@@ -44,29 +45,87 @@ module.exports = class CommentsService {
 
   /**
    * Adds a tag if it doesn't already exist on the comment.
+   * @throws if tag is already added to the comment
+   * @throws if tag name is not in ALLOWED_TAGS
+   * @param {String} id the id of the comment to tag
+   * @param {String} name the name of the tag to add
+   * @param {String} assigned_by the user id for the user who added the tag
    */
   static addTag(id, name, assigned_by) {
 
-    if (ALLOWED_TAGS.find((t) => t.name === name) == null) {
+    if (ALLOWED_TAGS.find((t) => t === name) == null) {
       return Promise.reject(new Error('tag not allowed'));
     }
 
-    return CommentModel.update({
+    const filter = {
       id,
-      tags: {
-        $ne: {
-          name
+      tags: {$not: {$elemMatch: {name}}}
+    };
+    const update = {
+      $push: {tags: {
+        name,
+        assigned_by,
+        created_at: new Date()
+      }}
+    };
+    return CommentModel.update(filter, update)
+      .then(({nModified}) => {
+        switch (nModified) {
+        case 0:
+
+            // either the tag was already there, or the comment doesn't exist with that id...
+          return this.findById(id)
```

Comment:
Can not use `this.findById` as all these functions do not provide context. This operation should not query the database at all, this is mixing of database queries which complicates the service interface. I do not expect that `addTag` would query for a comment, this would introduce a performance bottleneck that would be difficult to diagnose.